File: breifReview/Frontend/src/app/services/product.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, OnInit } from '@angular/core';
import { Observable } from 'rxjs';

const BASE_URL ="http://localhost:8082/";

@Injectable({
  providedIn: 'root'
})
export class ProductService {

  constructor(private http:HttpClient) { }
  productID : any;
  getAllProducts(){
    return this.http.get(BASE_URL+"api/getallproducts");
  }
  getProduct(product_code:string){
    return this.http.get(BASE_URL+"api/getproduct/"+product_code);
  }
  getProductById(product_id:string){
     return this.http.get(BASE_URL+"api/getproductbyid/"+product_id);
  }
  postReview(review:any):Observable<any>{
    console.log(review);
    return this.http.post<[]>(BASE_URL+"api/addreview",review); 
  }
  saveProductID(product_id:any){
    this.productID=product_id;
    return this.productID
  }
  getProductId(){
    return this.productID
  }
  getProductByBrand(brand:any){
    return this.http.get(BASE_URL+"api/getallproductsbybrand/"+brand);
  }
  getProductByCode(code:any){
    return this.http.get(BASE_URL+"api/getallproductsbycode/"+code);
  }
  getProductByName(name:any){
    return this.http.get(BASE_URL+"api/getallproductsbyname/"+name);
  }
  getReviews(product_id:string){
    return this.http.get(BASE_URL+"api/getreviews/"+product_id);
  }
  getProductReviewsCount(product_id:string){
    return this.http.get(BASE_URL+"api/getproductreviewscount/"+product_id);
  }
  getAverageRating(product_id:string){
    return this.http.get(BASE_URL+"api/getaveragerating/"+product_id);
  }
  saveProduct(product:any){
    return this.http.post<[]>(BASE_URL+"api/addproduct",product);
  }
  searchProducts(keywords:string[]){
    let params = new HttpParams();
    keywords.forEach((keyword, index) => {
      params = params.append('keywords', keyword);
    });
    return this.http.get(BASE_URL+"api/search/"+keywords);
  }

 
  }

File: breifReview/Backend/src/main/java/com/nagarro/communitybackend/services/ProductService.java
package com.nagarro.communitybackend.services;

import java.util.List;

import com.nagarro.communitybackend.dto.ProductDTO;
import com.nagarro.communitybackend.entities.Product;

public interface ProductService {
	
	Product saveProduct(ProductDTO product);  //saves the product in db

	List<Product> getAllProducts();   //fetches all products from db

	int getProductsCount();          //gets count of Products table

	Product getProductById(Long product_id);   //fetches product based upon its id

	Product getProduct(String product_code);   //fetches product based upon its code

	List<Product> getAllProductsByBrand(String brand);   //fetches list of products that matches the brand name

	List<Product> getAllProductsByCode(String code);     //fetches list of products that matches the code

	List<Product> getAllProductsByName(String name);     //fetches list of products that matches the product name
	
	List<Product> searchProducts(List<String> name);     //fetches list of products based upon the keywords entered.
	
}
